### PR TITLE
feat(net): use CompletionService to get ip more quickly

### PR DESCRIPTION
**Pull request by 317787106**

### CODE REVIEW

File: src/main/java/org/tron/p2p/utils/NetUtil.java [hunk 1]
```diff
@@ -195,4 +204,30 @@ public static InetSocketAddress parseInetSocketAddress(String para) {
...
+    try {
+      future = completionService.take();
+      result = future.get();
+    } catch (InterruptedException | ExecutionException e) {
+      //ignore
+    } finally {
+      executor.shutdownNow();
+    }
+
+    return result;
```
Comment (chengtx01) [inline src/main/java/org/tron/p2p/utils/NetUtil.java:230]: If there is an exception in obtaining the IP, should we consider returning a default value?

Comment (317787106) [reply src/main/java/org/tron/p2p/utils/NetUtil.java:230]: The logic is same as before. If it returns null, the result should be checked by upper business.

File: src/main/java/org/tron/p2p/base/Constant.java [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
+
 public class Constant {
 
   public static final int NODE_ID_LEN = 64;
-  public static final String AMAZONAWS_URL = "http://checkip.amazonaws.com";
-  public static final String IDENT_URL = "https://v6.ident.me";
+  public static final List<String> ipV4Urls = Arrays.asList(
+      "http://checkip.amazonaws.com", "https://ifconfig.me/", "https://4.ipw.cn/");
+  public static final List<String> ipV6Urls = Arrays.asList(
+      "https://v6.ident.me", "http://6.ipw.cn/");
```
[the remark this answers is not in the document]
Comment (317787106) [reply src/main/java/org/tron/p2p/base/Constant.java:12]: In fact, there are not many sites that provide this service. There is little necessary to config it manually.

File: src/test/java/org/tron/p2p/utils/NetUtilTest.java [hunk 3]
```diff
@@ -75,6 +78,22 @@ public void testExternalIp() {
...
+      Method method = NetUtil.class.getDeclaredMethod("getExternalIp", String.class);
+      method.setAccessible(true);
+      method.invoke(NetUtil.class, Constant.ipV4Urls.get(0));
+    } catch (Exception e) {
+      Assert.fail();
+    }
+    long t2 = System.currentTimeMillis();
+    NetUtil.getExternalIpV4();
+    long t3 = System.currentTimeMillis();
+    Assert.assertTrue(t2 - t1 >= t3 - t2);
```
Comment (halibobo1205) [inline src/test/java/org/tron/p2p/utils/NetUtilTest.java:94]: Results are randomized.

Comment (317787106) [reply src/test/java/org/tron/p2p/utils/NetUtilTest.java:94]: What do you mean "randomized"?

Comment (halibobo1205) [reply src/test/java/org/tron/p2p/utils/NetUtilTest.java:94]: `t2 - t1 >= t3 - t2` doesn't always hold true.

Comment (317787106) [reply src/test/java/org/tron/p2p/utils/NetUtilTest.java]: I have tried this many times, it works well though network is not avaiIable. But i comment this. Add test case of compare ip from different source.

**Review by chengtx01 — APPROVED** [review]

File: src/test/java/org/tron/p2p/utils/NetUtilTest.java [hunk 4]
```diff
@@ -75,6 +78,28 @@ public void testExternalIp() {
...
+    try {
+      Method method = NetUtil.class.getDeclaredMethod("getExternalIp", String.class);
+      method.setAccessible(true);
+      ip1 = (String) method.invoke(NetUtil.class, Constant.ipV4Urls.get(0));
+      ip2 = (String) method.invoke(NetUtil.class, Constant.ipV4Urls.get(1));
+      ip3 = (String) method.invoke(NetUtil.class, Constant.ipV4Urls.get(2));
+    } catch (Exception e) {
+      Assert.fail();
+    }
+    long t2 = System.currentTimeMillis();
```
Comment (halibobo1205) [inline src/test/java/org/tron/p2p/utils/NetUtilTest.java:94]: Please remove unused code.

File: src/test/java/org/tron/p2p/utils/NetUtilTest.java (hunk 3 above)
Comment (317787106) [reply src/test/java/org/tron/p2p/utils/NetUtilTest.java]: @chengtx01 @halibobo1205  If the node has multi outer extranet ip, querying from different urls may get different ip. Is there any problem?

File: src/test/java/org/tron/p2p/utils/NetUtilTest.java (hunk 4 above)
Comment (317787106) [reply src/test/java/org/tron/p2p/utils/NetUtilTest.java]: Already deleted.

**Review by halibobo1205 — APPROVED** [review]

**Pull request merged**